### PR TITLE
fix #20564 Align behavior of ONNX Frontend function Relu-6, 13, 14 wi…

### LINKED ISSUE #20564: [Good First Issue]: Align behavior of ONNX Frontend function Relu-6, 13, 14 with original framework
### Context

Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.
OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.
This task requires alignment between OpenVINO ONNX Frontend and original framework implementations of Relu for next list of opsets: [opset 6](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Relu-6),  [opset 13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Relu-13),  [opset 14](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Relu-14)
Necessary help will be provided by ONNX Fronted team.

### What needs to be done?

Operator details can be found in [ONNX Operators](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Relu)
More details can be found in ONNX Changelog: [opset 6](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Relu-6),  [opset 13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Relu-13),  [opset 14](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Relu-14)
1. Function already has a common implementation in OpenVINO. First of all, you need to review a documentation and prepare a table with differences between versions. It could be, for instance, a missing property, extended/reduced coverage of existing property, etc...
2. Copy existing implementation [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op) to make it aligned with original framework (extend or reduce coverage of a common implementation). Copy of modified implementation should be in a defined opset, or in opset 1 in case it implements oldest implementation. [Example of multi-opset operation](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp#L50).
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)
6. Check [Python xfailed tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/tests_python/test_backend.py) to find a test marked as a xfailed for added functionality. If any exist - remove corresponding lines and try to verify by using cmdline "python -m pytest -k name_of_test".
More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

### Example Pull Requests

_No response_

### Resources

- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [Contribution guide](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

### Contact points

@gkrivor 

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by Aryan8912**
close #20564 

first I understand the difference between Relu 6, 13, and 14  and implement the  here softmax.cpp 
then make a file relu6_13_14 and register a function [ops_bridge.cpp] 
…th original framework

### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Comment by p-wysocki** [comment]
build_jenkins

**Comment by p-wysocki** [comment]
Hi @Aryan8912, I see that there are no tests - could you please add them? That's the points 4, 5 and 6 in the Good First Issue description, "What needs to be done" section.

File: src/frontends/onnx/frontend/src/op/softmax.cpp [hunk 1]
```diff
@@ -47,6 +47,39 @@ OutputVector softmax(const Node& node) {
     return {result};
 }
 }  // namespace set_1
+namespace set_6 {
+
+OutputVector relu(const Node& node) {
```
Comment (p-wysocki) [inline src/frontends/onnx/frontend/src/op/softmax.cpp:52]: Should the ReLU be in softmax.cpp?

File: src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "openvino/core/deprecated.hpp"
+OPENVINO_SUPPRESS_DEPRECATED_START
```
Comment (p-wysocki) [inline src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp:8]: Should the new implementation be deprecated?

File: src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <memory>
+
+#include "default_opset.hpp"
+#include "ngraph/node.hpp"
+#include "onnx_import/core/node.hpp"
+
+namespace ngraph {
```
Comment (p-wysocki) [inline src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp:16]: Ngraph is deprecated API, could you please use the new one? I believe it's `ov::`.

**Review by p-wysocki — CHANGES_REQUESTED** [review]
Thank you for your contribution! Could you please add tests, fix the build and the formatter?

File: src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp (hunk 3 above)
Comment (Aryan8912) [reply src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp:16]: OK give me some time I will check some issues on my code and then resubmit the PR

File: src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp (hunk 2 above)
Comment (gkrivor) [reply src/frontends/onnx/frontend/src/op/relu_6_13_14.hpp:8]: until it contains ngraph - better to leave, easier for me to cleanup later

File: src/frontends/onnx/frontend/src/ops_bridge.cpp [hunk 4]
```diff
@@ -544,6 +544,10 @@ OperatorsBridge::OperatorsBridge() {
     REGISTER_OPERATOR("Upsample", 9, upsample);
 
     // custom ops
+    REGISTER_OPERATOR_WITH_DOMAIN(OPENVINO_ONNX_DOMAIN, "Relu", 6, set_6::relu);
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/ops_bridge.cpp:547]: you don't need to register it that way, use regular REGISTER_OPERATOR

**Comment by gkrivor** [comment]
Let me do some clarifications. We already have a file https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/relu.hpp with some "common implementation".

You may need to add a corresponding "relu.cpp" file. You need to move a function declaration there, and you need to leave only definition in a header. A moving all "inline" functions from headers is another part of refactoring.

And use the new relu.cpp to place declarations for new opsets there.

If you take a look in docs you will see, relu actually has version 1, 6, 13, 14, but actually it uses version 18... Version 18, as I saw previously, has just slightly different operations inside (uses operations from 18th opset). So we may not implement "version 18" at the moment.

As I can see, differences between versions are:
1 - 6 - lost a legacy attribute (which we actually didn't use (example: https://github.com/openvinotoolkit/openvino/blob/99d14a8b07a90f9328b208d8b881df84797b43fe/src/frontends/onnx/frontend/src/op/add.cpp#L18)
6-13 - extended list of supported input types (here is an example https://github.com/openvinotoolkit/openvino/blob/99d14a8b07a90f9328b208d8b881df84797b43fe/src/frontends/onnx/frontend/src/op/greater_or_equal.cpp#L22)
13-14 - extended list of supported input types

So, you will need to have 4 functions for 1-6, 6-13, 13-14, and 14+. Difference inside will be in an error messages depends on input types: tell user if it is using wrong combination of type-opset.

**Comment by Aryan8912** [comment]
please check my new update

**Comment by Aryan8912** [comment]
@gkrivor please check my PR update

**Review by p-wysocki — CHANGES_REQUESTED** [review]
Hello, I see that there's still `ngraph` namespace being used, as well as the deprecations. Could you please remove them?

**Comment by Aryan8912** [comment]
Sorry for the late reply sir, I just saw your comment, I am working on two issue right now so give some time I will update my code

**Comment by Aryan8912** [comment]
Hey sir, it is more than 5 days since I resubmitted my pull request, and there a no response, please if anything I missed please let me know.

**Comment by Aryan8912** [comment]
hey Sir, I have made a pull request , I am not able to understand what is the problem with this PR of mine. please help me out

**Comment by Aryan8912** [comment]
No please don't close pr

**Comment by Aryan8912** [comment]
Can you tell me what type of changes need to be made in the code?

File: src/frontends/onnx/frontend/src/op/add.cpp [hunk 5]
```diff
@@ -35,6 +35,18 @@ ov::OutputVector add(const Node& node) {
 }
 }  // namespace set_7
 
+namespace set_13 {
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/add.cpp:38]: Not sure I get why you did this... Please, check

File: src/frontends/onnx/frontend/src/op/greater_or_equal.cpp [hunk 6]
```diff
@@ -28,6 +28,46 @@ ov::OutputVector greater_or_equal(const Node& node) {
...
 }  // namespace set_1
 
+
+namespace onnx_import {
+namespace op {
+
+OutputVector relu(const Node& node) {
+    const auto opset_version = node.get_opset_version();
+
+    if (opset_version >= 1 && opset_version <= 6) {
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/greater_or_equal.cpp:38]: No need to do this here, it is done in ops_bridge.cpp logic when it is looking for correct implementation.

But maybe reasonable to re-use this approach for checking input types, depends on opset version. Could you try?

File: src/frontends/onnx/frontend/src/op/softmax.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/softmax.cpp:1]: Why did you remove it?

File: src/frontends/onnx/frontend/src/op/greater_or_equal.cpp
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/greater_or_equal.cpp:1]: Also, you don't need to do it in greater_or_equal.cpp. It has a separate file: https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/relu.hpp

Btw, what's the difference between versions? Am I right it is only in a list of supported types?

File: src/frontends/onnx/tests/onnx_import.in.cpp [hunk 7]
```diff
@@ -58,6 +58,46 @@ OPENVINO_TEST(${BACKEND_NAME}, onnx_test_test_case) {
     test_case.run();
 }
 
+OPENVINO_TEST(${BACKEND_NAME}, align_relu_opset_6) {
+    auto model = ngraph::onnx_import::import_onnx_model(
+        file_util::path_join(SERIALIZED_ZOO, "op/relu.hpp"));
```
Comment (gkrivor) [inline src/frontends/onnx/tests/onnx_import.in.cpp:63]: It is wrong, you should prepare *.prototxt files to use a converted *.onnx files in such place.

**Review by gkrivor — CHANGES_REQUESTED** [review]

**Comment by mlukasze** [comment]
hey @Aryan8912 
will you have time to finish this PR?
